### PR TITLE
[Merged by Bors] - Feat(RingTheory/Ideal/Maps): Golf & generalize `smul_top_eq_map`

### PR DESCRIPTION
**Pull request by Shamrock-Frost**
Golf & generalize `smul_top_eq_map`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Ideal/Maps.lean [hunk 1]
```diff
@@ -243,28 +243,25 @@ theorem le_comap_sup : comap f K ⊔ comap f L ≤ comap f (K ⊔ L) :=
...
+    r • N.restrictScalars R = (algebraMap R S r • N).restrictScalars R :=
+  SetLike.coe_injective (congrArg (· '' _) (funext (algebra_compatible_smul S r)))
+
+theorem smul_restrictScalars {R S M} [CommSemiring R] [CommSemiring S]
+    [Algebra R S] [AddCommMonoid M] [Module R M] [Module S M]
+    [IsScalarTower R S M] (I : Ideal R) (N : Submodule S M) :
+    I • N.restrictScalars R = (I.map (algebraMap R S) • N).restrictScalars R := by
+  simp_rw [map, Submodule.span_smul_eq, ← Submodule.coe_set_smul,
+    Submodule.set_smul_eq_iSup, element_smul_restrictScalars, iSup_image]
+  exact (_root_.map_iSup₂ (Submodule.restrictScalarsLatticeHom R S M) _).symm
```
Comment (erdOne) [inline Mathlib/RingTheory/Ideal/Maps.lean:258]: I think these should be flipped and maybe even be tagged `@[simp]`

Comment (erdOne) [reply Mathlib/RingTheory/Ideal/Maps.lean:258]: And the first lemma arguably belong somewhere else.

Comment (Shamrock-Frost) [reply Mathlib/RingTheory/Ideal/Maps.lean:258]: I was going for consistency with `smul_top_eq_map`. I agree that simplifying the other way makes sense but this seems backwards from the existing `simp`

Comment (erdOne) [reply Mathlib/RingTheory/Ideal/Maps.lean:258]: I think the simp direction is pushing `restrictScalars` inwards? For example `Submodule.restrictScalars_mul`

Comment (Shamrock-Frost) [reply Mathlib/RingTheory/Ideal/Maps.lean:258]: But then we have the simp rules `I • (⊤ : Submodule R S) → (I.map (algebraMap R S)).restrictScalars R` and `(I.map (algebraMap R S) • N).restrictScalars R → I • N.restrictScalars R`, which imo seem like they'e going in opposite directions?

Comment (erdOne) [reply Mathlib/RingTheory/Ideal/Maps.lean:258]: Oh I didn't know that was a simp lemma. Maybe just leave them as is for now and mark them later if needed. I still think flipping them is a good idea though.

Comment (Shamrock-Frost) [reply Mathlib/RingTheory/Ideal/Maps.lean]: Flipped

**Comment by erdOne** [comment]
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**